python/pyspark/worker.py

```diff
@@ -1514,10 +1514,293 @@ def _remove_partition_by_exprs(self, arg: Any) -> Any:
+            """
+            Create a new instance that wraps the provided Arrow UDTF with partitioning
+            logic.
+
+            Parameters
+            ----------
+            create_udtf: function
+                Function that creates a new instance of the Arrow UDTF to invoke.
+            partition_child_indexes: list
+                Zero-based indexes of input-table columns that contain projected
+                partitioning expressions.
+            """
+            self._create_udtf: Callable = create_udtf
+            self._udtf = create_udtf()
+            self._partition_child_indexes: list = partition_child_indexes
+            # Track last partition key from previous batch
+            self._last_partition_key: Optional[Tuple[Any, ...]] = None
+            self._eval_raised_skip_rest_of_input_table: bool = False
+
+        def eval(self, *args, **kwargs) -> Iterator:
+            """Handle partitioning logic for Arrow UDTFs that receive RecordBatch objects."""
+            import pyarrow as pa
+
+            # Get the original batch with partition columns
+            original_batch = self._get_table_arg(list(args) + list(kwargs.values()))
+            if not isinstance(original_batch, pa.RecordBatch):
+                # Arrow UDTFs with PARTITION BY must have a TABLE argument that
+                # results in a PyArrow RecordBatch
+                raise PySparkRuntimeError(
+                    errorClass="INVALID_ARROW_UDTF_TABLE_ARGUMENT",
+                    messageParameters={
+                        "actual_type": str(type(original_batch))
+                        if original_batch is not None
+                        else "None"
+                    },
+                )
+
+            # Remove partition columns to get the filtered arguments
+            filtered_args = [self._remove_partition_by_exprs(arg) for arg in args]
+            filtered_kwargs = {
+                key: self._remove_partition_by_exprs(value) for (key, value) in kwargs.items()
+            }
+
+            # Get the filtered RecordBatch (without partition columns)
+            filtered_batch = self._get_table_arg(filtered_args + list(filtered_kwargs.values()))
+
+            # Process the RecordBatch by partitions
+            yield from self._process_arrow_batch_by_partitions(
+                original_batch, filtered_batch, filtered_args, filtered_kwargs
+            )
+
+        def _process_arrow_batch_by_partitions(
+            self, original_batch, filtered_batch, filtered_args, filtered_kwargs
+        ) -> Iterator:
+            """Process an Arrow RecordBatch by splitting it into partitions.
+
+            Since Catalyst guarantees that rows with the same partition key are contiguous,
+            we can use efficient boundary detection instead of group_by.
+
+            Handles two scenarios:
+            1. Multiple partitions within a single RecordBatch (using boundary detection)
```

Comment:
just curious why we allow multiple partitions within a single batch?
I am working on other python plans, and two different groups cannot be in the same batch.